File: src/SOP8/Egovframework/Base/base-account-bll/src/main/java/egovframework/base/account/bll/process/OptionManager.java
package egovframework.base.account.bll.process;

import java.sql.Connection;
import java.sql.SQLException;
import java.util.*;

import egovframework.base.dal.account.OptionDAO;
import egovframework.base.generic.Holder;
import egovframework.base.model.account.Option;
import egovframework.base.account.ibll.models.UserOption;
import egovframework.base.account.ibll.request.RequestOption;
import egovframework.base.account.ibll.response.ResponseOption;
import egovframework.base.response.MessageResult;
import egovframework.base.account.bll.resource.ErrorMessage;
import egovframework.base.response.resource.ID;
import egovframework.base.dal.IDatabase;

public class OptionManager {
    public static ResponseOption getOption(RequestOption data, OptionDAO optionDao) {
        ResponseOption result = new ResponseOption();
        Holder<String> errorMessage = new Holder<>();

        String strCondition = String.format("%s = %d", Option.Fields.user_sn, data.getUserNo());

        if (data.getCategory() != null && !data.getCategory().isEmpty()) {
            if (data.getSubCategory() != null && !data.getSubCategory().isEmpty()) {
                strCondition += String.format(" and %s = '%s' and %s = '%s'",
                        Option.Fields.optn_cl, data.getCategory(),
                        Option.Fields.optn_sclas, data.getSubCategory());
            } else {
                strCondition += String.format(" and %s = '%s'",
                        Option.Fields.optn_cl, data.getCategory());
            }
        }

        strCondition += String.format(" order by %s, %s, %s",
                Option.Fields.optn_cl,
                Option.Fields.optn_sclas,
                Option.Fields.optn_indx);

        List<Option> options = optionDao.selectOptions(null, strCondition, null, errorMessage);

        if (options == null) {
            result.setSuccess(false);
            result.setMessage(ID.get(ErrorMessage.class, "failToReadOption").value());
            return result;
        }

        result.setSuccess(true);
        result.setUserNo(data.getUserNo());
        result.setOptions(new ArrayList<>());
        result.getOptions().addAll(toUserOptions(options));
        return result;
    }

    private static List<UserOption> toUserOptions(List<Option> options) {
        Map<String, UserOption> dicUserOptions = new HashMap<>();

        for (Option option : options) {
            String strKey = option.getOptn_cl();
            if (option.getOptn_sclas() != null)
                strKey += "_" + option.getOptn_sclas();

            UserOption userOption = dicUserOptions.get(strKey);
            if (userOption == null) {
                userOption = new UserOption();
                userOption.setCategory(option.getOptn_cl());
                userOption.setSubCategory(option.getOptn_sclas());
                userOption.setValues(new ArrayList<>());
                dicUserOptions.put(strKey, userOption);
            }

            userOption.getValues().add(option.getOptn_value());
        }

        return new ArrayList<>(dicUserOptions.values());
    }

    public static MessageResult saveOption(int userNo, List<UserOption> options, OptionDAO optionDao, IDatabase database) {
        Holder<String> errorMessage = new Holder<>();

        Connection connection = null;

        try {
            connection = database.getConnection();

            // Begin Transaction
            // 트랜잭션 수동제어
            connection.setAutoCommit(false);

            for (UserOption option : options) {
                String strCondition = String.format("%s = %d", Option.Fields.user_sn, userNo);

                if (option.getCategory() != null) {
                    if (option.getSubCategory() != null) {
                        strCondition += String.format(" and %s = '%s' and %s = '%s'",
                                Option.Fields.optn_cl, option.getCategory(),
                                Option.Fields.optn_sclas, option.getSubCategory());
                    } else {
                        strCondition += String.format(" and %s = '%s'",
                                Option.Fields.optn_cl, option.getCategory());
                    }
                }

                if (!optionDao.deleteOption(strCondition, errorMessage)) {
                    connection.rollback();
                    return new MessageResult(false, errorMessage.value);
                }
            }

            List<Option> userOptions = new ArrayList<>();

            for (UserOption option : options) {
                if (option.getValues() == null) continue;

                for (int i = 0; i < option.getValues().size(); i++) {
                    Option userOption = new Option();
                    userOption.setUser_sn(userNo);
                    userOption.setOptn_cl(option.getCategory());
                    userOption.setOptn_sclas(option.getSubCategory());
                    userOption.setOptn_indx(i + 1);
                    userOption.setOptn_value(option.getValues().get(i));
                    userOptions.add(userOption);
                }
            }

            if (!userOptions.isEmpty()) {
                if (!optionDao.insertOptions(userOptions, errorMessage)) {
                    connection.rollback();
                    return new MessageResult(false, errorMessage.value);
                }
            }
        }
        catch (Exception e) {
            if (connection != null) {
                try {
                    connection.rollback();
                }
                catch (SQLException ex) {}
            }

            e.printStackTrace();
            return new MessageResult(false, e.getMessage());
        }
        finally {
            if (connection != null) {
                try {
                    // commit
                    // 트랜잭션 자동제어(rollback을 호출했더라도 반드시 호출되어야 한다.)
                    connection.setAutoCommit(true);
                    connection.close();
                }
                catch (SQLException ex) {
                    return new MessageResult(false, "데이터베이스의 트랜잭션을 정상적으로 종료할 수 없습니다.");
                }
            }
        }

        return new MessageResult(true, "");
    }
}

File: src/SOP8/Base/Account/Base.Account.BLL/Process/OptionManager.cs
using dnsDapperDBUtil.DataAccessLayer.IDAL;
using System.Collections.Generic;
using Base.Account.IBLL.Response;
using Base.Account.IBLL.Request;
using Base.Account.IBLL.Models;
using Base.Model.Account;
using Response.Resource;
using Response;

namespace Base.Account.BLL.Process
{
    using Resource;

    class OptionManager
    {
        public static ResponseOption GetOption(RequestOption data, IDataManager dataManager)
        {
            ResponseOption result = new ResponseOption();

            string strErrorMessage = null;
            string strCondition = string.Format("{0} = {1}", Option.Fields.user_sn, data.UserNo);

            if (data.Category != null && data.Category.Length > 0)
            {
                if (data.SubCategory != null && data.SubCategory.Length > 0)
                    strCondition += string.Format(" and {0} = '{1}' and {2} = '{3}'",
                        Option.Fields.optn_cl, data.Category,
                        Option.Fields.optn_sclas, data.SubCategory);
                else
                    strCondition += string.Format(" and {0} = '{1}'",
                        Option.Fields.optn_cl, data.Category);
            }

            strCondition += string.Format(" order by {0}, {1}, {2}", Option.Fields.optn_cl, Option.Fields.optn_sclas, Option.Fields.optn_indx);

            IEnumerable<Option> options = dataManager.GetSelect().Select<Option>(strCondition, out strErrorMessage);

            if (options == null)
            {
                result.Success = false;
                result.Message = ID.Get<ErrorMessage>("failToReadOption").Value();
                return result;
            }

            result.Success = true;
            result.UserNo = data.UserNo;
            result.Options = new List<UserOption>();
            result.Options.AddRange(ToUserOptions(options));
            return result;
        }

        private static IEnumerable<UserOption> ToUserOptions(IEnumerable<Option> options)
        {
            Dictionary<string, UserOption> dicUserOptions = new Dictionary<string, UserOption>();
            UserOption userOption;

            foreach (Option option in options)
            {
                string strKey = option.optn_cl;

                if (option.optn_sclas != null)
                    strKey += "_" + option.optn_sclas;

                if (dicUserOptions.TryGetValue(strKey, out userOption) == false)
                {
                    userOption = new UserOption();
                    userOption.Category = option.optn_cl;
                    userOption.SubCategory = option.optn_sclas;

                    dicUserOptions[strKey] = userOption;
                }

                userOption.Values.Add(option.optn_value);
            }

            return dicUserOptions.Values;
        }

        public static MessageResult SaveOption(int userNo, List<UserOption> options, IDataManager dataManager)
        {
            string strErrorMessage = null;

            dataManager = dataManager.Clone();

            if (dataManager.BeginBatch(out strErrorMessage) == false)
                return new MessageResult(false, "시스템 데이터베이스의 트랜잭션을 시작할 수 없습니다.");

            foreach (UserOption option in options)
            {
                // 기존 옵션 삭제
                string strCondition = string.Format("{0} = {1}", Option.Fields.user_sn, userNo);

                if (option.Category != null)
                {
                    if (option.SubCategory != null)
                        strCondition += string.Format(" and {0} = '{1}' and {2} = '{3}'", Option.Fields.optn_cl, option.Category, Option.Fields.optn_sclas, option.SubCategory);
                    else
                        strCondition += string.Format(" and {0} = '{1}'", Option.Fields.optn_cl, option.Category);
                }

                if (dataManager.GetDelete().Delete<Option>(strCondition, out strErrorMessage) == false)
                {
                    string strTemp;
                    dataManager.BatchRollback(out strTemp);
                    return new MessageResult(false, strErrorMessage);
                }
            }

            List<Option> userOptions = new List<Option>();

            foreach (UserOption option in options)
            {
                if (option.Values == null)
                    continue;

                int valueCount = option.Values.Count;

                for (int i=0;i<valueCount;i++)
                {
                    Option userOption = new Option();

                    userOption.user_sn = userNo;
                    userOption.optn_cl = option.Category;
                    userOption.optn_sclas = option.SubCategory;
                    userOption.optn_indx = i + 1;
                    userOption.optn_value = option.Values[i];

                    userOptions.Add(userOption);
                }
            }

            if (userOptions.Count > 0)
            {
                if (dataManager.GetCreate().Insert<Option>(userOptions, out strErrorMessage) == false)
                {
                    string strTemp;
                    dataManager.BatchRollback(out strTemp);
                    return new MessageResult(false, strErrorMessage);
                }
            }

            if (dataManager.BatchCommit(out strErrorMessage) == false)
            {
                string strTemp;
                dataManager.BatchRollback(out strTemp);
                return new MessageResult(false, "시스템 데이터베이스의 트랜잭션을 종료할 수 없습니다.");
            }

            return new MessageResult(true, "");
        }
    }
}
